File: src/KM.Api/GameDump/GameDumpContracts.cs
// SPDX-License-Identifier: GPL-3.0-only

using KM.Api.Diagnostics;
using KM.Api.Projects;

namespace KM.Api.GameDump;

public enum GameDumpCategoryKindDto
{
    Table,
    Text,
    Raw,
}

public enum GameDumpFormatDto
{
    Tsv,
    Csv,
    Json,
    TsvAndJson,
    Txt,
    TxtAndJson,
    Raw,
    RawAndJson,
}

public sealed record LoadGameDumpWorkflowRequest(ProjectPathsDto Paths);

public sealed record RunGameDumpRequest(
    ProjectPathsDto Paths,
    string DestinationFolder,
    IReadOnlyList<GameDumpSelectionDto> Selections,
    string? ProducerVersion = null);

public sealed record GameDumpSelectionDto(
    string CategoryId,
    GameDumpFormatDto Format,
    IReadOnlyList<string>? LanguageCodes = null);

public sealed record GameDumpCategoryDto(
    string Id,
    string Label,
    string Description,
    GameDumpCategoryKindDto Kind,
    IReadOnlyList<GameDumpFormatDto> Formats,
    GameDumpFormatDto DefaultFormat,
    bool IsAvailable,
    IReadOnlyList<ApiDiagnostic> Diagnostics)
{
    public GameDumpCategoryLanguageOptionsDto? LanguageOptions { get; init; }
}

public sealed record GameDumpLanguageOptionDto(
    string Code,
    string Label);

public sealed record GameDumpCategoryLanguageOptionsDto(
    IReadOnlyList<GameDumpLanguageOptionDto> Options,
    IReadOnlyList<string> DefaultLanguageCodes,
    bool SupportsAllLanguages);

public sealed record GameDumpWorkflowDto(
    IReadOnlyList<GameDumpCategoryDto> Categories,
    IReadOnlyList<ApiDiagnostic> Diagnostics);

public sealed record GameDumpWrittenFileDto(
    string CategoryId,
    string RelativePath,
    long SizeBytes);

public sealed record GameDumpResultDto(
    string DestinationFolder,
    IReadOnlyList<GameDumpWrittenFileDto> WrittenFiles,
    IReadOnlyList<ApiDiagnostic> Diagnostics,
    bool Succeeded);

public sealed record LoadGameDumpWorkflowResponse(GameDumpWorkflowDto Workflow);

public sealed record RunGameDumpResponse(GameDumpResultDto Result);

File: apps/desktop/src/bridge/gameDumpContracts.ts
/* SPDX-License-Identifier: GPL-3.0-only */

import { z } from 'zod';
import { apiDiagnosticSchema, projectPathsSchema } from './contracts';

export const gameDumpCategoryKindSchema = z.enum(['table', 'text', 'raw']);
export const gameDumpFormatSchema = z.enum([
  'tsv',
  'csv',
  'json',
  'tsvAndJson',
  'txt',
  'txtAndJson',
  'raw',
  'rawAndJson'
]);

export const gameDumpSelectionSchema = z.strictObject({
  categoryId: z.string(),
  format: gameDumpFormatSchema
});

export const loadGameDumpWorkflowRequestSchema = z.strictObject({
  paths: projectPathsSchema
});

export const runGameDumpRequestSchema = z.strictObject({
  destinationFolder: z.string(),
  paths: projectPathsSchema,
  selections: z.array(gameDumpSelectionSchema)
});

export const gameDumpCategorySchema = z.strictObject({
  defaultFormat: gameDumpFormatSchema,
  description: z.string(),
  diagnostics: z.array(apiDiagnosticSchema),
  formats: z.array(gameDumpFormatSchema),
  id: z.string(),
  isAvailable: z.boolean(),
  kind: gameDumpCategoryKindSchema,
  label: z.string()
});

export const gameDumpWorkflowSchema = z.strictObject({
  categories: z.array(gameDumpCategorySchema),
  diagnostics: z.array(apiDiagnosticSchema)
});

export const gameDumpWrittenFileSchema = z.strictObject({
  categoryId: z.string(),
  relativePath: z.string(),
  sizeBytes: z.number().int().nonnegative()
});

export const gameDumpResultSchema = z.strictObject({
  destinationFolder: z.string(),
  diagnostics: z.array(apiDiagnosticSchema),
  succeeded: z.boolean(),
  writtenFiles: z.array(gameDumpWrittenFileSchema)
});

export const loadGameDumpWorkflowResponseSchema = z.strictObject({
  workflow: gameDumpWorkflowSchema
});

export const runGameDumpResponseSchema = z.strictObject({
  result: gameDumpResultSchema
});

export type GameDumpCategory = z.infer<typeof gameDumpCategorySchema>;
export type GameDumpCategoryKind = z.infer<typeof gameDumpCategoryKindSchema>;
export type GameDumpFormat = z.infer<typeof gameDumpFormatSchema>;
export type GameDumpResult = z.infer<typeof gameDumpResultSchema>;
export type GameDumpSelection = z.infer<typeof gameDumpSelectionSchema>;
export type GameDumpWorkflow = z.infer<typeof gameDumpWorkflowSchema>;
export type LoadGameDumpWorkflowRequest = z.infer<typeof loadGameDumpWorkflowRequestSchema>;
export type LoadGameDumpWorkflowResponse = z.infer<typeof loadGameDumpWorkflowResponseSchema>;
export type RunGameDumpRequest = z.infer<typeof runGameDumpRequestSchema>;
export type RunGameDumpResponse = z.infer<typeof runGameDumpResponseSchema>;
